bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/BoschSslUtil.java

```diff
@@ -0,0 +1,185 @@
+import org.openhab.core.id.InstanceUUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * SSL context utility.
+ *
+ * @author Gerd Zanker - Initial contribution
+ */
+@NonNullByDefault
+public class BoschSslUtil {
+
+    private static final String OSS_OPENHAB_BINDING = "oss_openhab_binding";
+
+    private final Logger logger = LoggerFactory.getLogger(BoschSslUtil.class);
+
+    private String keystorePath;
+    private String keystorePassword;
+
+    public static String getBoschSHCId() {
+        return OSS_OPENHAB_BINDING + "_" + InstanceUUID.get();
+    }
+
+    public BoschSslUtil(String keystorePassword) {
+        this.keystorePath = Paths.get(OpenHAB.getUserDataFolder(), "etc", getBoschSHCId() + ".jks").toString();
+        this.keystorePassword = keystorePassword;
+    }
+
+    public SslContextFactory getSslContextFactory() throws PairingFailedException {
+        // Instantiate and configure the SslContextFactory
+        SslContextFactory sslContextFactory = new SslContextFactory(true); // Accept all certificates
+
+        // during pairing the cert from this keystore is accessed by HTTP client via name
+        sslContextFactory.setKeyStore(getKeyStoreAndCreateIfNecessary());
+
+        // Keystore for managing the keys that have been used to pair with the SHC
+        // https://www.eclipse.org/jetty/javadoc/9.4.12.v20180830/org/eclipse/jetty/util/ssl/SslContextFactory.html
+        sslContextFactory.setKeyStorePath(keystorePath);
+        sslContextFactory.setKeyStorePassword(keystorePassword);
+
+        // Bosch is using a self signed certificate
+        sslContextFactory.setTrustAll(true);
+        sslContextFactory.setValidateCerts(false);
+        sslContextFactory.setValidatePeerCerts(false);
+        sslContextFactory.setEndpointIdentificationAlgorithm(null);
+
+        return sslContextFactory;
+    }
+
+    public KeyStore getKeyStoreAndCreateIfNecessary() throws PairingFailedException {
+        try {
+            File file = new File(keystorePath);
+            if (!file.exists()) {
+                // create new keystore
+                logger.info("Creating new keystore {} because it doesn't exist.", keystorePath);
+                return createKeyStore(keystorePath, keystorePassword);
+            } else {
+                // load keystore as a first check
+                KeyStore keyStore = KeyStore.getInstance("JKS");
+                keyStore.load(new FileInputStream(file), keystorePassword.toCharArray());
+                logger.debug("Using existing keystore {}", keystorePath);
+                return keyStore;
+            }
+        } catch (OperatorCreationException | GeneralSecurityException | IOException e) {
+            logger.debug("Exception during keystore creation {}", e.getMessage());
+            throw new PairingFailedException("Can not create or load keystore file: " + keystorePath
+                    + ". Check path, write access and JKS content.");
```

Comment:
Always include the cause exception when throwing another exception.
```suggestion
            throw new PairingFailedException("Can not create or load keystore file: " + keystorePath
                    + ". Check path, write access and JKS content.", e);
```